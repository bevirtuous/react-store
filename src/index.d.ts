import React from 'react';

export interface Action<T = any> {
  type: T
}

export interface AnyAction extends Action {
  // Allows any extra properties to be defined in an action.
  [extraProps: string]: any
}

export interface State {
  [prop: string]: any;
}

export interface Dispatch {
  (action: AnyAction, ...extraArgs: any[]): State
}

export interface Reducer {
  (state: State, action: AnyAction): State
}

export interface CombinedReducers {
  [reducer: string]: Reducer;
}

export interface GetState {
  (): State;
}

export interface ReactStore {
  dispatch: Dispatch;
  getState: GetState;
  action?: AnyAction;
  __state?: State;
}

export interface StoreData {
  store: ReactStore;
}

export interface StoreProps {
  children: React.ReactNode;
  reducer: Reducer;
  initialState?: State;
  middlewares?: Function[];
}

/**
 * Creates a store that holds the state tree.
 * The only way to change the data in the store is to call `dispatch()` on it.
 *
 * There should only be a single store in your app. To specify how different
 * parts of the state tree respond to actions, you may combine several reducers
 * into a single reducer function by using `combineReducers`.
 */
export function Store(props: StoreProps): JSX.Element;

export interface MapStateToProps<P = any, R = any> {
  (state: State, ownProps: P): R;
}

export interface MapDispatchToProps<P = any, R = any> {
  (dispatch: Dispatch, ownProps: P): R;
}

export type Selector<P, R> = (state: State, props?: P, ...args: any[]) => R;

/**
 * Combines an object of many reducers to one new higher order reducer.
 */
export function combineReducers(reducers: CombinedReducers): Reducer;

/**
* Dispatches an **action** from anywhere in your codebase.
*/
export function dispatch(action: AnyAction): void;

export function select<P, R>(selector: Selector<P, R>, props?: P): R;

/**
 * Returns the global application store's **dispatch** function.
 */
export function useDispatch(): Dispatch;

/**
 * Takes a **selector** and returns its output value.
 */
export function useSelector<P, R>(selector: Selector<P, R>, props: P): R;

/**
 * Returns an instance of the store.
 */
export function useStore(): ReactStore;
