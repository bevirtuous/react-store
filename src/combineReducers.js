/**
 * Combines an object of many reducers to one new higher order reducer.
 * @param {Object} reducers An object of multiple reducers.
 * @returns {Function} A new higher order reducer that unifies all the others.
 */
export function combineReducers(reducers = {}) {
  return (state = {}, action) => (
    Object.keys(reducers).reduce((nextState, key) => {
      // eslint-disable-next-line no-param-reassign
      nextState[key] = reducers[key](state[key], action);
      return nextState;
    }, {})
  );
}
