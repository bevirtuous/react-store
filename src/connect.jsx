import React from 'react';
import StoreContext from './context';

export default (mapStateToProps, mapDispatchToProps) => (
  Component => props => (
    <StoreContext.Consumer>
      {({ dispatch, getState }) => (
        <Component
          {...props}
          {...(mapStateToProps(getState(), props) || {})}
          {...(mapDispatchToProps(dispatch, props) || {})}
        />
      )}
    </StoreContext.Consumer>
  )
);
