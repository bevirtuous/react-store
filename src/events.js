import EventEmitter from 'events';

export const STORE_DISPATCH = 'STORE_DISPATCH';

export const emitter = new EventEmitter();
