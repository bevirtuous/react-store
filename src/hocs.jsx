import React from 'react';
import { StoreContext } from './context';

/**
 * Injects the store's dispatch function into the component props.
 * @param {Function} Component The original component.
 * @returns {Function}
 */
export function withDispatch(Component) {
  return function WithDispatch(props) {
    return (
      <StoreContext.Consumer>
        {({ dispatch }) => <Component {...props} dispatch={dispatch} />}
      </StoreContext.Consumer>
    );
  };
}

/**
 * Injects the store properties into the component props.
 * @param {Function} Component The original component.
 * @returns {Function}
 */
export function withStore(Component) {
  return function WithStore(props) {
    return (
      <StoreContext.Consumer>
        {store => <Component {...props} {...store} />}
      </StoreContext.Consumer>
    );
  };
}
