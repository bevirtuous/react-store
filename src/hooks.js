import { useContext } from 'react';
import StoreContext from './context';

/**
 * Returns the global application store's **dispatch** function.
 * @returns {Function} The global dispatch function.
 */
export const useDispatch = () => {
  const { dispatch } = useContext(StoreContext);

  return dispatch;
};

/**
 * Takes a **selector** and returns its output value.
 * @param {Function} selector The selector to exectue.
 * @param {Object} [props] Some component props to pass to the selector.
 * @returns {*} The output value of the selector.
 */
export const useSelector = (selector, props = undefined) => {
  const { getState } = useContext(StoreContext);

  return selector(getState(), props);
};

/**
 * Returns an instance of the store.
 * @returns {Object} The application store.
 */
export const useStore = () => {
  const store = useContext(StoreContext);

  return store;
};
