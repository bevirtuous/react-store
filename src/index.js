export { Store } from './Store';
export { useDispatch, useSelector, useStore } from './hooks';
export { withDispatch, withStore } from './hocs';
export { connect } from './connect';
export { combineReducers } from './combineReducers';
export { dispatch, select } from './helpers';
