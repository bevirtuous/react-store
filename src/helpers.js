import { emitter, STORE_DISPATCH } from './events';
import { storeData } from './Store';

/**
 * Dispatches action **action** from anywhere in your codebase.
 * @param {Object} action A plain object representing "what changed". An action must have
 * a `type` property which may not be `undefined`. It is a good idea to use
 * string constants for action types.
 */
export const dispatch = (action) => {
  if (typeof action === 'undefined') {
    throw new Error('The dispatch helper can only be called with a valid action.');
  }
  emitter.emit(STORE_DISPATCH, action);
};

export const select = (selector, props) => {
  if (typeof selector === 'undefined') {
    throw new Error('The select helper can only be called with a valid selector');
  }

  return selector(storeData.store.getState(), props);
};
