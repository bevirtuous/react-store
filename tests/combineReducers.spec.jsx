import { combineReducers } from '../src/combineReducers';

describe('combineReducers', () => {
  it('should return an empty object if no reducers are passed', () => {
    const reducer = combineReducers();
    const state = reducer({});
    expect(state).toEqual({});
    expect(state).toMatchSnapshot();
  });

  it('should return the correct state if called', () => {
    const reducer = combineReducers({ test: () => ({ foo: 'bar' }) });
    const state = reducer({});
    expect(state).toEqual({ test: { foo: 'bar' } });
    expect(state).toMatchSnapshot();
  });
});
