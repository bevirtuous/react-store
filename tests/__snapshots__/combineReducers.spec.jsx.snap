// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`combineReducers should return an empty object if no reducers are passed 1`] = `Object {}`;

exports[`combineReducers should return the correct state if called 1`] = `
Object {
  "test": Object {
    "foo": "bar",
  },
}
`;
